From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Wed, 7 May 2025 20:08:13 +0000
Subject: [PATCH 107/240] NVIDIA: SAUCE: tegra_hv: clear out memory while
 unreserving ivc queue

Bug 5433066

Change-Id: I143f3b5ab508ff6d973a908f13357ab0edfc465c
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3451317
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: svc-mobile-coverity <svc-mobile-coverity@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
(cherry picked from commit ec3b32edfa1e6bb85d30734a9cbc7e7863241410)

diff --git a/drivers/firmware/tegra/ivc.c b/drivers/firmware/tegra/ivc.c
index 8c9aff9804c0..75e50a7da4a0 100644
--- a/drivers/firmware/tegra/ivc.c
+++ b/drivers/firmware/tegra/ivc.c
@@ -643,6 +643,13 @@ static inline void *iosys_map_get_vaddr(const struct iosys_map *map)
 	return map->vaddr;
 }
 
+void tegra_ivc_clean_queue_data(struct tegra_ivc *ivc)
+{
+	iosys_map_memset(&ivc->rx.map, sizeof(struct tegra_ivc_header), 0, ivc->num_frames * ivc->frame_size);
+	iosys_map_memset(&ivc->tx.map, sizeof(struct tegra_ivc_header), 0, ivc->num_frames * ivc->frame_size);
+}
+EXPORT_SYMBOL(tegra_ivc_clean_queue_data);
+
 int tegra_ivc_init(struct tegra_ivc *ivc, struct device *peer, const struct iosys_map *rx,
 		   dma_addr_t rx_phys, const struct iosys_map *tx, dma_addr_t tx_phys,
 		   unsigned int num_frames, size_t frame_size,
diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index 371f4ea4c58b..77b5534b357d 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -895,6 +895,7 @@ int tegra_hv_ivc_unreserve(struct tegra_hv_ivc_cookie *ivck)
 			ivc_release_irq(ivc);
 		ivc->cookie_ops = NULL;
 		ivc->reserved = 0;
+		tegra_ivc_clean_queue_data(&ivc->ivc);
 		ret = 0;
 	} else {
 		ret = -EINVAL;
diff --git a/include/soc/tegra/ivc.h b/include/soc/tegra/ivc.h
index be45d5f5adea..736590774114 100644
--- a/include/soc/tegra/ivc.h
+++ b/include/soc/tegra/ivc.h
@@ -29,6 +29,15 @@ struct tegra_ivc {
 	size_t frame_size;
 };
 
+/**
+ * tegra_ivc_clean_queue_data - Zeroed out the IVC queue data
+ * @ivc		pointer of the IVC channel
+ *
+ * Zeroed out the IVC queue data .
+ *
+ */
+void tegra_ivc_clean_queue_data(struct tegra_ivc *ivc);
+
 /**
  * tegra_ivc_read_get_next_frame - Peek at the next frame to receive
  * @ivc		pointer of the IVC channel
-- 
2.55.0

